// @sap/cds-dk/lib/init/template/destinations/index.js
const { join } = require('path')
const { readProject } = require('../../util/projectReader')
const { mergeYAML } = require('../../util/merge')
const { parseMd5File, writeMd5File } = require('../../util/md5')

module.exports = class DestinationsTemplate extends require('../templateBase') {

    async canRun() {
        const projectDescriptor = await readProject(this.options)
        const { hasMta, hasHelm } = projectDescriptor.cap
        if (!hasHelm && hasMta) {
            throw `'cds add destinations' is not available for Cloud Foundry yet`
        }
        return true
    }

    static hasFacet(env) {
        const kinds = { odata: 1, 'odata-v2': 1, 'odata-v4': 1, rest: 1 }
        const mtxServices = { 'cds.xt.SaasProvisioningService': 1, 'cds.xt.DeploymentService': 1, 'cds.xt.ModelProviderService': 1, 'cds.xt.ExtensibilityService': 1 }
        const _fromMtx = service => service in mtxServices
        return Object.entries(env.requires).some(([service, require]) => !_fromMtx(service) && !kinds[service] && kinds[require.kind])
    }

    async run() {
        await this.runDependentMerging()
    }

    async runDependentMerging(context = {}) {
        const projectDescriptor = await readProject(this.options)
        const { hasHelm, hasDestination, hasHtml5Repo } = projectDescriptor.cap

        if (hasHelm) {
            //const env = await this.getEnv()
            const binding = hasDestination && { srv: { bindings: { destination: { serviceInstanceName: 'destination' }}}}
            const html5Binding = hasHtml5Repo && { 'html5-apps-deployer': { bindings: { destination: { serviceInstanceName: 'destination' }}}}

            //in case facet is being added to the already exisiting charts folder
            let isIndependentCommand = false
            if(Object.keys(context).length == 0){
                isIndependentCommand = true
                context.oldTrackingData = context.newTrackingData = await parseMd5File(join('chart', '.cds-add-helm-files.md5'))
            }
            await mergeYAML(
                join('chart', 'values.yaml'),
                {
                    ...binding,
                    ...html5Binding,
                    destination: {
                        serviceOfferingName: 'destination',
                        servicePlanName: 'lite',
                        parameters: {
                            version: '1.0.0',
                            ...(hasHtml5Repo) && { HTML5Runtime_enabled: true }
                        }
                    }
                }
            )

            // copy service instance chart if it is not present
            const HelmTemplate = require(`../helm`);
            const helmTemplate = new HelmTemplate(this.generator, context.oldTrackingData, context.newTrackingData);
            isIndependentCommand = isIndependentCommand | await helmTemplate.addSubChart('service-instance');

            // add dependency entry in chart.yaml
            await helmTemplate.addDependency('service-instance', 'destination');

            if(isIndependentCommand) await writeMd5File(join('chart', '.cds-add-helm-files.md5'), context.newTrackingData)
        }
    }
}
